tests/providers/openlineage/plugins/test_listener.py

```diff
@@ -33,10 +33,20 @@
 from airflow.providers.openlineage.plugins.listener import OpenLineageListener
 from airflow.providers.openlineage.utils.selective_enable import disable_lineage, enable_lineage
 from airflow.utils.state import State
+from tests.test_utils.compat import AIRFLOW_V_2_10_PLUS
 from tests.test_utils.config import conf_vars
 
 pytestmark = pytest.mark.db_test
 
+EXPECTED_TRY_NUMBER_1 = 1 if AIRFLOW_V_2_10_PLUS else 0
```

Comment:
@mobuchowski  @kacpermuda @dstandish -> this was the easiest way I found to make the same tests work before and after the try_num change. I am not sure if this is the **right** or **easy** way to do it, and what's the reasoning behind having those tests, but certainly it looks like with the `try_num` changes, the values that open-lineage listener will receive will change - the tests now work for both pre- and post- `try_num` change.

I basically have two questions:

* is it ok to keep it this way? (@dstandish ?) 
* does it not mean that open-lineage integration for Airlfow 2.10 is heavily impacted by the `try_num` change (@mobuchowski , @kacpermuda ) ?

Reply:
Sorry for the late response, a lot going on today.

> does it not mean that open-lineage integration for Airlfow 2.10 is heavily impacted by the try_num change?

Yes, but that was already taken car of in this [PR](https://github.com/apache/airflow/pull/39336/files#diff-776b432a685e8cc2a2ec0432aab9d9b22ca4e79a247debf0967a0a5802af09f2R49) i think.

> is it ok to keep it this way?

@mobuchowski @JDarDagran wdyt?
I think some tests will not be necessary and some new are needed if the try_num will remain unchanged during the entire execution. I can prepare a separate PR for it on Monday, that will update the OL tests. I'm not sure yet how i could  recreate in tests the entire flow of the taskinstance and be sure that the try_num is not changing, but I'll figure something out.

More info:
In OpenLineage we are generating a unique identifier for each task run using dag_id, task_id, execution_date and try_number. We needed to make sure that those run_id are deterministic and constant across the OL run (so that all listener methods receive task_instance with specific try_num), and that's what these tests are supposed to do. As I could not fully mock and recreate the task execution flow together with the exact moment when the try_num is increased, i did my best to simply test whether or not the OL code responds well to current hard-coded moment when the try_num was increased.

Reply:
> Yes, but that was already taken car of in this [PR](https://github.com/apache/airflow/pull/39336/files#diff-776b432a685e8cc2a2ec0432aab9d9b22ca4e79a247debf0967a0a5802af09f2R49) i think.

Well. But openlineage provider should (or not?) handle try_num before and after the change? Or the part where openlineage retrieves and uses try_num is part of the Airflow core? Because if it is in the provider then both - installing new openlineage provider in Airflow 2.9 wll not work, and installing old openlineage provider in Airlfow 2.10 (or 2.9.2 if we cherry-pick the change to 2.9) will not work either I am afraid.

I thin - since this test is in the provider, the fact that we had to modify it for 2.9.1 is an indication that things will break.

Reply:
cc: @dstandish

Reply:
> Well. But openlineage provider should (or not?) handle try_num before and after the change? 

Maybe I am missing something, but I think that what's happening now after #39336 has been merged. The tests requires some adjusting, but the core of OL works fine. The PR already modifies the OL listener to handle `try_num` depending on Airflow version (`_get_try_number_success()` function). The only difference between previous and new `try_num` (for OL provider use case) is that `try_num` is not increased before `on_task_instance_success` is called, so inside that method we don't have to subtract 1 from `try_num` to get the same value as in `on_task_instance_failed` and `on_task_instance_running`. From Airflow 2.10 (or 2.9.2 if cherry picked) we can use `try_num` without any modifications in whole OL listener.

As to the versions:
- For Airflow >= 2.10 (or 2.9.2 if cherry picked) the OL provider < 1.7.2 (or 1.8.0 depending on release) will not work properly, as it will not contain this `try_num` adjustment. We should maybe provide a max Airflow version for previous releases of OL provider or a min provider version for Airflow version starting 2.10 (or 2.9.2 if cherry picked).
- For all airflow versions (>= 2.7 when OL provider was introduced) the OL provider >= 1.7.2 (or 1.8.0 depending on release) should work just fine, imho, as there is an adjustment for Airflow version (`_get_try_number_success()` function).

Reply:
@kacpermuda is right - we had to work around `try_number` handling before, now we don't - but still have to correct previous behavior for Airflow < 2.10.

Reply:
Yes. I think "old OL provider"  + "Airlfow 2.9" work fine. New OL provider + Airflow 2.10 as well. What is problematic is what happens when we mix them. 

> We should maybe provide a max Airflow version for previous releases of OL provider.

This is not possible. Release is immutable once released.

> or a min provider version for Airflow version starting 2.10 (or 2.9.2 if cherry picked).

We are not going to cherry-pick that change, I think - seeing how it interacts with OL (@dstandish ?) . We cannot really have mininimum version specified in dependencies in a "hard" way (optional dependencies like `openlineage` extra can have limits but they are not enforced, they are only used at installation time temporarily). 

Those checks will have to be done at runtime. I think what we will have to do is to:

1) have runtime check in airflow to see if openlineage is installed and if so, check it's version is > 1.8.0
2) in openlinage provider have compatibility code to check if airflow version is < 2.10 and apply the old try_num adjustments in this case

Reply:
@potiuk I know that we can specify min airflow version for a specific provider version, but is there any way we can specify min version of OL provider for specific Airflow versions? We could then specify that Airflow >= 2.10 will not accept OL provider < 1.8.0. Of course Airflow should not require OL, just disallow Ol versions smaller than 1.8.0.

Reply:
> @potiuk I know that we can specify min airflow version for a specific provider version, but is there any way we can specify min version of OL provider for specific Airflow versions? We could then specify that Airflow >= 2.10 will not accept OL provider < 1.8.0. Of course Airflow should not require OL, just disallow Ol versions smaller than 1.8.0.

As explained above - yes and no. We can specify it in extra:

```
openlineage: ["apache-airflow-provider-openlineage>=1.8.0"]
```

But iff the dependency is optional, it only has temporary effect during installation and ONLY if the extra is actually used. It's not enforceable - for example you will be able to install `apache-airflow-providers-openlineage==1.7.0` on `apache-airflow==2.10.0` regardless what `openlinege` extra says.

So yes - we can (and probably should) specify min version of openlineage provider, but also, if we really want to enforce it, it has to be additionally checked at runtime. Many other libraries do very similar checks, this is where you sometimes see a warning when starting airflow `you have this and that version of x library but you have xxxx library that is not compatible` - apache.beam and few other libraries do that for their extras.

It should be easy to do: 

```
openlineage_version = importlib.metadata.version("apache-airflow-providers-openlineage") 
if openlineage_version and openlineage_version < .....:
     log.warning(,,,)
```

Reply:
(we can also error out if openlineage is enabled and version is < expected)

Reply:
Yes, i did not refresh the page and did not see your first comment 😄 I'll prepare some kind of check for that, as you described

Reply:
Any particular place i could create this runtime check at? I could not find this beam one you mentioned. Any directions here about the desired location would be appreciated.

Reply:
Beam does it internally - but in this case I think it should be done in a general airflow `settings.py` (somewhere in initialize()` ) there are few other generic checks, and this one could be triggered by having open lineage enabled in configuration.

Reply:
#39609 Will add the runtime check. And when it comes to those OL tests, i will improve them as a separate PR i guess. For now, if they work it's fine.

Reply:
Makes sense.